Refs #35192 -- Added possibility to override BoundField at Form level or Field level (#18266)
django/django: docs/ref/forms/api.txt

``use_required_attribute=False`` to avoid incorrect browser validation when
adding and deleting forms from a formset.

.. attribute:: Form.bound_field_class

.. versionadded:: 5.2

Specify a :class:`~django.forms.BoundField` class to be used on the form. For
example::

    from django import forms

    class MyForm(forms.Form):

>         bound_field_class = CustomBoundField

Reviewer: Can you please define in this example how `CustomBoundField` looks like?

Reviewer 2: The shortening of the example was asked for earlier; I think we should instead shorten the example even more and only point readers to the `Customizing BoundField` section.

``use_required_attribute=False`` to avoid incorrect browser validation when
adding and deleting forms from a formset.

.. attribute:: Form.bound_field_class

Allows you set a different :class:`~django.forms.BoundField` class to use for the form. This
allows to override, for instance, :meth:`~django.forms.BoundField.label_tag`/

> :meth:`~django.forms.BoundField.legend_tag` to use in templates.

Reviewer 2: Maybe:

"As an example, this allows overriding `label_tag` or `legend_tag` which are used by the default form rendering in templates:"

Also, further down the form field API docs mention customizing bound fields; the introduction of the "Customizing BoundField" paragraph should possibly be reworded to explicitly mention that this is a way to customize the BoundField class per-field? It's probably obvious from the context but I wanted to point it out.

Author: > Also, further down the form field API docs mention customizing bound fields

You mean [this section](https://github.com/django/django/blob/main/docs/ref/forms/api.txt#L1147)?

Reviewer 2: No, further below: 

https://github.com/django/django/blob/c2c7dbb2f88ce8f0ef6d48a61b93866c9926349a/docs/ref/forms/api.txt#L1444-L1483

Author: Done.

Author: Ok done. Is that ok?

Reviewer 2: I think it's great, thank you.

``use_required_attribute=False`` to avoid incorrect browser validation when
adding and deleting forms from a formset.

> .. attribute:: Form.bound_field_class

Reviewer: This addition is inside a subsection named "Configuring form elements' HTML ``id`` attributes and ``<label>`` tags". I'm not sure this content is relevant for that? Could you please share your rationale for why this was added here? Thanks.

Reviewer: Maybe a good location would be inside the "Styling required or erroneous form rows"?

Reviewer 2: Mentioning it in the form rows section is a great idea; I added a link from there to the `Customizing BoundField` section.

.. class:: BoundField

>     Used to display HTML or access attributes for a single field of a

Reviewer: Could you please keep the original 3-space indentation? that maps to the `.. ` characters.

Reviewer 3: This was already discussed [here](https://github.com/django/django/pull/18266#discussion_r1836612171)
I agreed this is consistent with the rest of the docs

>     Used to display HTML or access attributes for a single field of a
>     :class:`Form` instance.
> 
>     The ``__str__()`` method of this object displays the HTML for this field.
> 
>     You can use :attr:`.Form.bound_field_class` and
>     :attr:`.Field.bound_field_class` to use a different ``BoundField`` class
>     per-form or per-field.

Reviewer 3: ```suggestion
   Used to display HTML or access attributes for a single field of a
   :class:`Form` instance.

   The ``__str__()`` method of this object displays the HTML for this field.

   You can use :attr:`.Form.bound_field_class` and
   :attr:`.Field.bound_field_class` to use a different ``BoundField`` class
   per-form or per-field.
```
(nit) this has been indented slightly off

Author: What? 3 spaces instead of 4? I don't understand: it's not consistant with the rest of the documentation, is it?

Reviewer 3: Ah you're right, it was indented wrong previously and you have fixed it :+1:

>     See :ref:`custom-boundfield` for examples of overriding a ``BoundField``.

Reviewer 3: ```suggestion
   See :ref:`custom-boundfield` for examples of overriding a ``BoundField``.
```

   Used to display HTML or access attributes for a single field of a
   :class:`Form` instance.

   You can use :attr:`~django.forms.Form.bound_field_class` and
   :attr:`~django.forms.Field.bound_field_class` to use a different
   ``BoundField`` class per-form or per-field.

>    The ``__str__()`` method of this object displays the HTML for this field.

Reviewer 2: This has been applied, except for the versionchanged suggestion. I think if people click through to the `bound_field_class` description they will see that this is a new feature.

.. method:: BoundField.render(template_name=None, context=None, renderer=None)

>     The render method is called by ``as_field_group``. All arguments are

Reviewer: Despite this blank space deleted is a nice cleanup, it should not be part of this commit. It could be a previous commit or a different PR.

Author: I wonder whether those are not corrected by the pre-commit hooks.

Reviewer: I don't understand exactly your question, but to be more precise: these empty space cleanups should not be part of this PR/commit, they should be a different commit or PR altogether. Thanks!

Author: since PyCharm is constantly autocorrecting this because of `trim_trailing_whitespace` = true in `.editorconfg` I can't prevent this from being part of the diff. I will always endup having this correction as soon as I open the file. It's just a pain to prevent. It does not seem to be worth the cost to leave this correction out of the diff so is it ok to leave it in?

Reviewer: Thank you @christophehenry for the details. Sadly,  is not OK to leave those in, it's against policy. You are welcome to do a separated PR with those fixes, or a previous commit in this same PR. Ping me explicitly and I'll merge ASAP.

Author: [Done](https://github.com/django/django/pull/18856).

        >>> print(bound_form["subject"].value())
        hi

.. _custom-boundfield:

Customizing ``BoundField``
==========================

A form can override :attr:`.Form.bound_field_class` and a form field can take
an optional :attr:`.Field.bound_field_class` constructor argument.

For example, you may need to access some additional information about a form
field in a template and using a subclass of :class:`~django.forms.Field` isn't
sufficient.

If you have a ``GPSCoordinatesField``, for example, and want to be able to

> access additional information about the coordinates in a template, this could

Reviewer 3: I'm sorry I didn't realize that `get_bound_field` is documented in an old release notes and overriding it is also covered in tests, given that I think this should start with:
```
There are multiple ways to use a custom :class:`.BoundField`:

* A project can define a project-wise :class:`.BoundField` class using
  :attr:`.BaseRenderer.bound_field_class` and :setting:`FORM_RENDERER`.

* A form can override :attr:`.Form.bound_field_class` or use the constructor
  argument.

* A custom form field can override :attr:`.Field.bound_field_class` or use the
  constructor argument.

* A custom form field can override :meth:`.Field.get_bound_field()`:

.. method:: Field.get_bound_field(form, field_name)

    Takes an instance of :class:`~django.forms.Form` and the name of the field.
    The return value will be used when accessing the field in a template. Most
    likely it will be an instance of a subclass of :class:`.BoundField`.

You may want to use a custom :class:`.BoundField` if you need to access some
additional information about a form field in a template and using a subclass of
:class:`~django.forms.Field` isn't sufficient.

For example, if you have a ``GPSCoordinatesField``, for example, and want to be
able to access additional information about the coordinates in a template, this
could be implemented as follows::
```
(sorry for the back and forth)

        >>> print(bound_form["subject"].value())
        hi

.. _custom-boundfield:

Customizing ``BoundField``
==========================

There are multiple ways to use a custom :class:`.BoundField`:

> * A project can define a project-wise :class:`.BoundField` class using

Reviewer: Did you mean project-wide here?
```suggestion
* A project can define a project-wide :class:`.BoundField` class using
```

Reviewer: @christophehenry This wasn't fixed, which is unfortunate. I'll make a push with this and a few other small tweaks, so before any future changes, please re-fetch my changes once I complete the review.

Reviewer 2: Fixed now. (Sorry, I hope we don't duplicate work here.)

  :attr:`.BaseRenderer.bound_field_class` and :setting:`FORM_RENDERER`.

> * A form can override :attr:`.Form.bound_field_class` or use the constructor
>   argument.

Reviewer: ```suggestion
* A form can override :attr:`.Form.bound_field_class` or set the
  ``bound_field_class`` argument in the constructor.
```

> * A custom form field can override :attr:`.Field.bound_field_class` or use the
>   constructor argument.

Reviewer: ```suggestion
* A custom form field can override :attr:`.Field.bound_field_class` or set the
  ``bound_field_class`` argument in the constructor.
```

> * A custom form field can override :meth:`.Field.get_bound_field()`:

Reviewer: Following @smithdc1 comment, perhaps we could find a middle ground saying:
```suggestion
* Lastly, a custom form field can override :meth:`.Field.get_bound_field()`
  though any of the previous options is preferred:
```

.. method:: Field.get_bound_field(form, field_name)

    Takes an instance of :class:`~django.forms.Form` and the name of the field.
    The return value will be used when accessing the field in a template. Most

>     likely it will be an instance of a subclass of :class:`.BoundField`.

Reviewer: Is this diff line necessary?

Reviewer 3: I suggested this change ```:class:`.BoundField`.``` is being used in this section as to me it renders better and is easier to read in the short form

You may want to use a custom :class:`.BoundField` if you need to access some
additional information about a form field in a template and using a subclass of
:class:`~django.forms.Field` isn't sufficient.

> For example, if you have a ``GPSCoordinatesField``, for example, and want to be

Reviewer: This needs rewording as there are two repeated "for example".

        >>> print(bound_form["subject"].value())
        hi

.. _custom-boundfield:

Customizing ``BoundField``
==========================

There are multiple ways to use a custom :class:`.BoundField`:

* A project can define a project-wide :class:`.BoundField` class using
  :attr:`.BaseRenderer.bound_field_class` and :setting:`FORM_RENDERER`.

* A form can override :attr:`.Form.bound_field_class` or set the constructor
  ``bound_field_class`` argument in the constructor.

* A custom form field can override :attr:`.Field.bound_field_class` or set the
  ``bound_field_class`` argument in the constructor.

* Lastly, a custom form field can override :meth:`.Field.get_bound_field()`
  though any of the previous options is preferred:

> .. method:: Field.get_bound_field(form, field_name)

Author: So what's the decision here? Should this be moved in `fields.txt`?

template and using a subclass of :class:`~django.forms.Field` isn't

> sufficient, consider also customizing :class:`~django.forms.BoundField`.

Reviewer 3: I mean something like this:
```diff
 Customizing ``BoundField``
 ==========================
 
-If you need to access some additional information about a form field in a
-template and using a subclass of :class:`~django.forms.Field` isn't
-sufficient, consider also customizing :class:`~django.forms.BoundField`.
-
 A form can override :attr:`.Form.bound_field_class` and a form field can take
 an optional :attr:`.Field.bound_field_class` constructor argument.
 
+For example, you may need to access some additional information about a form
+field in a template and using a subclass of :class:`~django.forms.Field` isn't
+sufficient.
+
 If you have a ``GPSCoordinatesField``, for example, and want to be able to
 access additional information about the coordinates in a template, this could
 be implemented as follows::
@@ -1500,6 +1500,29 @@ be implemented as follows::
 Now you can access the country in a template with
 ``{{ form.coordinates.country }}``.
 
+You may also want to customize the default form field template rendering. For
+example, you can override :meth:`.BoundField.label_tag` to add a custom class::
+
+
+   class StyledLabelBoundField(BoundField):
+       def label_tag(self, contents=None, attrs=None, label_suffix=None, tag=None):
+           attrs = attrs or {}
+           attrs["class"] = "wide"
+           return super().label_tag(contents, attrs, label_suffix, tag)
+
+
+   class UserForm(forms.Form):
+       bound_field_class = StyledLabelBoundField
+       name = CharField()
+
+This would update the default form rendering:
+
+.. code-block:: pycon
+
+   >>> f = UserForm()
+   >>> print(f["name"].label_tag)
+   <label for="id_name" class="wide">Name:</label>
+
 A project can define a project-wise ``BoundField`` class using
 :attr:`.BaseRenderer.bound_field_class` and :setting:`FORM_RENDERER`.
```

A project can define a project-wise ``BoundField`` using
:attr:`~django.forms.renderers.BaseRenderer.bound_field_class` and

> :setting:`FORM_RENDERER`.

Reviewer 2: It is at the top of the list now.

template and using a subclass of :class:`~django.forms.Field` isn't
sufficient, consider also customizing :class:`~django.forms.BoundField`.

A custom form can override :attr:`~django.forms.Form.bound_field_class` and

> a custom form field takes an optionnal
> :attr:`~django.forms.Field.bound_field_class` constructor argument.
>  Additionnally, it can override ``get_bound_field()``:

Reviewer 2: ```suggestion
a custom form field takes an optional
:attr:`~django.forms.Field.bound_field_class` constructor argument.
 Additionally, it can override ``get_bound_field()``:
```

Author: Yes, the CI pointed me that. I just fixed this.

> Additionally, it can override ``get_bound_field()``:

Reviewer 4: I am not sure that we should still recommend overriding `get_bound_field()`. Let's just point folk at the new way.

Reviewer 2: This has been reworded in the meantime; `get_bound_field` is still mentioned, but all other ways are recommended first. I think the current version reads well.

    class GPSCoordinatesField(Field):
        bound_field_class = GPSCoordinatesBoundField

Now you can access the country in a template with
``{{ form.coordinates.country }}``.

You may also want to customize the default form field template rendering. For
example, you can override :meth:`.BoundField.label_tag` to add a custom class::

   class StyledLabelBoundField(BoundField):
       def label_tag(self, contents=None, attrs=None, label_suffix=None, tag=None):
           attrs = attrs or {}
           attrs["class"] = "wide"
           return super().label_tag(contents, attrs, label_suffix, tag)

   class UserForm(forms.Form):
       bound_field_class = StyledLabelBoundField
       name = CharField()

This would update the default form rendering:

.. code-block:: pycon

   >>> f = UserForm()
   >>> print(f["name"].label_tag)
   <label for="id_name" class="wide">Name:</label>

A project can define a project-wise ``BoundField`` class using

> :attr:`.BaseRenderer.bound_field_class` and :setting:`FORM_RENDERER`.

Reviewer: I'm applying this suggestion since it hasn't been applied yet.
